vms/platformvm/state/state.go

```diff
@@ -1944,3 +2010,50 @@ func (s *state) writeMetadata() error {
 	}
 	return nil
 }
+
+// TODO: Remove after next hard-fork when everybody has run this.
+func (s *state) populateBlockHeightIndex() error {
+	blk, _, err := s.GetStatelessBlock(s.lastAccepted)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve last accepted block from local disk: %w", err)
+	}
+
+	heightKey := database.PackUInt64(blk.Height())
+	if _, err := database.GetID(s.blockIDDB, heightKey); err == nil {
+		// If the last accepted block is in [s.blockIDDB], no backfilling is required.
```

Comment:
I think this is wrong. Even if the process is pretty fast, there is always the chance that it gets interrupted at some point.
If the lastAccepted block height index is written before any other height, this check won't make the reindex fill the gap at the next node run and the index will stay incomplete.
I think the proper way to reindex is what we did in proposervm:
1. Have a canonical way to iterate blocks (picking current block parent instead of the db iterator)
2. Checkpoint at times to have a efficient way to handle previous incomplete reindex ops. Say you checkpoint every 100 blocks. At restart if there is a checkpoint check 100 blocks below that checkpointed height: if any is missing you resume reindexing till genesis; if there are all there index is complete (just checkpoint was not properly persisted)

Reply:
1cc8adfc118e37e15dc13e5b471b5f860db5e127 I added an explicit "skip the last accepted block" in the for loop. Adding checkpointing seems like overkill because it is so fast

Reply:
I am not sure I agree that checkpoints are an overkill. It may happen rarely, but if it does I don't think the user would have any way to check the index is complete nor other ways to fix it except wipe out the whole db and recreate the node.
I think we should use proposerVM reindexing as a blueprint for how to build the height index.
Also, similarly to what we did with proposerVM, we could drop the whole rebuild logic at the next hard-fork, acting as if height index has been always there for our whole community.
@gyuho, @StephenButtolph @danlaine  thoughts?

Reply:
> the user would have any way to check the index is complete nor other ways to fix it except wipe out the whole db and recreate the node

I may be misunderstanding something here but I don't think this is true after the most recent changes to the code. The `lastAcceptedBlock` is only written to `blockIDDb` at the end of the height index generation so we should be able to use it to tell if the index is complete.

Reply:
You're right!

Reply:
This is all being written to a versiondb - so everything is committed atomically (I removed the special case handling for  the last accepted block)